twilight-gateway-queue/src/lib.rs

```diff
@@ -1,153 +1,56 @@
-    /// session.
-    fn request<'a>(&'a self, shard_id: [u32; 2]) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>>;
-}
-
-/// A local, in-process implementation of a [`Queue`] which manages the
-/// connection attempts of one or more shards.
-///
-/// The queue will take incoming requests and then queue them, releasing one of
-/// the requests every 6 seconds. The queue is necessary because there's a
-/// ratelimit on how often shards can initiate sessions.
-///
-/// Handling shard queues usually won't need to be manually handled due to the
-/// gateway having built-in queueing when managing multiple shards.
-///
-/// # When not to use this
-///
-/// This queue implementation is "local", meaning it's intended to be used if
-/// you manage shards only in this process. If you run shards in multiple
-/// different processes (do you utilize microservices a lot?), then you **must
-/// not** use this implementation. Shards across multiple processes may
-/// create new sessions at the same time, which is bad.
-///
-/// It should also not be used for very large sharding, for that the
-/// [`LargeBotQueue`] can be used.
-///
-/// If you can't use this, look into an alternative implementation of the
-/// [`Queue`], such as the [`gateway-queue`] broker.
-///
-/// [`gateway-queue`]: https://github.com/twilight-rs/gateway-queue
-#[derive(Clone, Debug)]
-pub struct LocalQueue(UnboundedSender<Sender<()>>);
-
-impl Default for LocalQueue {
-    fn default() -> Self {
-        Self::new()
-    }
-}
-
-impl LocalQueue {
-    /// Creates a new local queue.
-    pub fn new() -> Self {
-        let (tx, rx) = unbounded_channel();
-
-        tokio::spawn(waiter(rx));
-
-        Self(tx)
-    }
-}
-
-async fn waiter(mut rx: UnboundedReceiver<Sender<()>>) {
-    const DUR: Duration = Duration::from_secs(6);
-    while let Some(req) = rx.recv().await {
-        if let Err(source) = req.send(()) {
-            tracing::warn!("skipping, send failed: {source:?}");
-        } else {
-            sleep(DUR).await;
-        }
-    }
-}
-
-impl Queue for LocalQueue {
-    /// Request to be able to identify with the gateway. This will place this
-    /// request behind all other requests, and the returned future will resolve
-    /// once the request has been completed.
-    fn request(&'_ self, [id, total]: [u32; 2]) -> Pin<Box<dyn Future<Output = ()> + Send + '_>> {
-        Box::pin(async move {
-            let (tx, rx) = oneshot::channel();
-
-            if let Err(source) = self.0.send(tx) {
-                tracing::warn!("skipping, send failed: {source:?}");
-                return;
-            }
-
-            tracing::info!("shard {id}/{total} waiting for allowance");
-
-            _ = rx.await;
-        })
-    }
+    /// Send `()` to signal the shard to proceed. Note that shards may have
+    /// dropped the receiver prior.
+    ///
+    /// Closing the channel should causes the shard to requeue.
+    fn enqueue(&self, id: u32) -> oneshot::Receiver<()>;
```

Comment:
In twilight we have prior art to hiding implementation details, even if it involves API-stable dependencies such as tokio, see for example [http-ratelimiting](https://github.com/twilight-rs/twilight/blob/main/twilight-http-ratelimiting/src/ticket.rs#L75). For consistency, we should make a wrapper newtype around `oneshot::Receiver<()>`.

Reply:
Yeah I can see the argument, but I'm not sure there's any benefit except for "hiding" implementation details. http-ratelimiting's wrapping allows naming the types (instead of, for example, `Receiver<Sender<Headers>>`) but that's not applicable here…